### PR TITLE
Add ems id to the person and primary persons table

### PR DESCRIPTION
**Pull request by roseeichelmann**
## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/14592

## Testing
**URL to test:** 
Local DB

**Steps to test:**
1. Checkout this branch, start up your local DB and apply the migrations and metadata in this branch
```bash
hasura migrate apply --envfile .env.local
hasura metadata apply --envfile .env.local
``` 
2. Nothing else really to test here, no UI changes. You can test going to the `ems__incidents` table, copying the `id` of one of the records, and then add this id to the `ems_id` field of a record in `atd_txdot_person`. Then you can see the `ems_incident` relationship for that record.

---
#### Ship list
- [ ] Check migrations for any conflicts with latest migrations in master branch
- [ ] Code reviewed
- [x] Product manager approved

### CODE REVIEW

**Review by johnclary — COMMENTED** [review]
i know foreign key constraints are not common in this DB, but is there any reason not to use them in this case?

**Comment by roseeichelmann** [comment]
@johnclary good question. my logic for not adding foreign key constraints is that not all person records will have an ems incident record associated with it. conversely, not all ems records will have person records associated with them bc it will be up to the vz team to manually fill in the `ems_id` column for a person record in order to manually link them.

**Comment by mddilley** [comment]
Taking a quick peek here, and I see what John is saying. I think that a foreign key constraint would be a good idea to make sure an entered value has a corresponding record with that id in the EMS incidents table. Having a null foreign key should be valid, but I could be missing a detail. Also raises the question of what the unique id column for `ems__incidents` is - looks like `incident_number`?

Unless the VZ team wants the ability to enter an `ems_id` that doesn't exist in the EMS incidents table yet. That would change things.

**Comment by roseeichelmann** [comment]
@mddilley Ooo i guess i misconstrued foreign keys and forgot they could be null. I'll edit to add a foreign key constraint!

I think the unique identifier would be just the `id` column, according to the column comment for `incident_number` "Unique identifier for the Incident. Note that this value may not be unique to records in this dataset, as there may be multiple patient care records for a single incident".

**Comment by johnclary** [comment]
👍 this is all sounding good to me. one consideration is how this might affect whatever ETLs you run on the `ems__incidents` table. e.g. if you ever truncate/delete records from that table, the foreign key constraint can add a layer of complication.

**Comment by mddilley** [comment]
Good call, John! @roseeichelmann I'm reading the issue again, and I think it could be worth figuring out exactly which EMS column the VZ team expects to enter in this column before proceeding. Maybe we can dedicate some time to this in refinement tomorrow (if we are all there). Great catch on the `incident_number` - `pcr_key` could be an alternative since it is noted as unique per row.

**Comment by johnclary** [comment]
in case this is helpful wrt the foreign key column to use: it's worth keeping in mind that the linking field we present in the UI does not have to the be the value stored in the foreign key column. e.g. if we have our own PK/ID sequence in the EMS table, and we have another EMS-provided column which uniquely identifies a record, it's probably right to use our own ID column as the foreign key because we have full control over that column and it can't be mucked up by some upstream issue. i do not say that with a ton of experience dealing with this, though!

**Comment by johnclary** [comment]
just want to make sure @frankhereford has put eyes on this

**Comment by frankhereford** [comment]
> just want to make sure @frankhereford has put eyes on this

Thanks @johnclary - I actually have a review going that I had not completed, and I'll do that right now. 🙏

File: atd-vzd/migrations/default/1700863342765_set_fk_public_atd_txdot_primaryperson_ems_id/up.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+alter table "public"."atd_txdot_primaryperson"
+  add constraint "atd_txdot_primaryperson_ems_id_fkey"
+  foreign key ("ems_id")
+  references "public"."ems__incidents"
+  ("id") on update restrict on delete set null;
```
Comment (frankhereford) [inline atd-vzd/migrations/default/1700863342765_set_fk_public_atd_txdot_primaryperson_ems_id/up.sql:5]: There's nothing easy to me about visualizing what the `on update restrict on delete set null` does, but I think you have it set just like 100% perfectly here. I read this as, you can't update a record's ID in the `ems__incidents` table if you have primary persons pointing at that ID and if you delete the record out of `ems__incidents`, you end up setting the FK in the primary persons to null.

I did verify how the EMS ETL works, and I think this is all going to work together, but [please take a look](https://github.com/cityofaustin/atd-vz-data/blob/master/atd-etl/afd_ems_import/ems_incident_upload.py#L265-L277) if you don't mind and make sure you agree with me.

**Review by frankhereford — APPROVED** [review]
Rose, I think this looks great. I've done my best to look into how the EMS ETL will interact with the referential sanity you're enforcing with your constraints. I think you're right on the money.

I offered a little observation about some terminology in your comments that felt new to me, but those suggestions couldn't be less important and certainly don't block this getting approved. 

Thanks for doing this!!

**Comment by frankhereford** [comment]
@johnclary @roseeichelmann I just read other's reviews and comments after submitting mine. I try to do mine with as blank a mind as possible so I don't end up focusing on what other have focused on.

I think that the issues John has raised very tightly align to when we discussed this in a recent meeting -- maybe last refinement? I can't recall. But - WRT the ETL, I think we're 👍 here after reviewing it, and I'm all for constraints whenever we can use them -- they are hard to write and review, but oh do they pay out in the long run enforcing referential sanity.

**Review by mddilley — APPROVED** [review]
I tested this locally by going up and down the migrations, testing the relationship in a GraphQL query in the Hasura console, and updating rows in the person and primary persons tables with valid and invalid foreign keys. It all works great! 🚀🚢

**Review by patrickm02L — APPROVED** [review]
Send it 🥏

File: atd-vzd/migrations/default/1700162755001_alter_table_public_atd_txdot_primaryperson_add_column_ems_id/up.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+alter table "public"."atd_txdot_primaryperson" add column "ems_id" integer
+ null;
+
+COMMENT ON COLUMN public.atd_txdot_primaryperson.ems_id IS 'This field has an object relationship with the "id" field of the ems__incidents table.';
```
[the remark this answers is not in the document]
Comment (roseeichelmann) [reply atd-vzd/migrations/default/1700162755001_alter_table_public_atd_txdot_primaryperson_add_column_ems_id/up.sql:4]: Yeah I was using the wording from the hasura relationship types @mddilley linked. I can see how its confusing though so I'll update to just mention foreign keys!

**Pull request merged**